### ISSUE TITLE
mla_decode_fwd (ASM sparse-MLA decode) returns NaN nondeterministically with num_kv_splits=16

### ISSUE THREAD

**Issue by yhl-amd** [ROCm/aiter#4350]
## Environment

- GPU: AMD Instinct MI350-series (gfx950)
- ROCm/HIP: 7.2.53211
- torch: 2.10.0+rocm7.2.4.git3d3aa833
- aiter: commit 95abba475 (main)
- Python: 3.12.3

## Summary

`aiter.mla.mla_decode_fwd`, called with realistic DeepSeek-V3.2/GLM-5.2 sparse-MLA decode dimensions (B=1, H=16, D=576, topk=2048 selected KV positions out of a 65536-token context), returns an entirely-NaN output tensor when `num_kv_splits=16`. `num_kv_splits` values of 1, 8, and 32 on the identical inputs are all correct (match a plain torch reference to ~2e-4 relative error). `num_kv_splits=16` is the split count real inference stacks use in production (e.g. ATOM passes `num_kv_splits=16` whenever the model is not using a "segmented" MLA cache layout — the common/default case).

## Reproduction

Attached script: `aiter_issue_repro.py`. A sweep of `num_kv_splits ∈ {1, 8, 16, 32}` on fixed inputs reproduced NaN at exactly 16 in 4/4 attempts on our box:

```
  num_kv_splits=  1: max_abs_diff_vs_ref=0.00021 nan_count=0/8192
  num_kv_splits=  8: max_abs_diff_vs_ref=0.00020 nan_count=0/8192
  num_kv_splits= 16: max_abs_diff_vs_ref=nan     nan_count=8192/8192
  num_kv_splits= 32: max_abs_diff_vs_ref=0.00019 nan_count=0/8192
```

Looping `num_kv_splits=16` alone 100x on byte-identical inputs occasionally shows a mixed pass/fail pattern (once observed 50/100 NaN, other times 0/100) — evidence this is a **race condition** (timing/scheduling-dependent), not a fixed wrong-answer-for-this-input bug. Most likely a missing synchronization/memory-fence in the `num_kv_splits=16` split-K reduction stage, where a worker's partial-sum write isn't yet visible when the reducing block reads it — the race's outcome appears sensitive to GPU state/occupancy left over from preceding kernel activity (calling `num_kv_splits=16` as the very first kernel launch in a fresh process did not reproduce it for us).

## Impact

In production serving (ATOM + GLM-5.2-MXFP4/FP8 on TP=8), this intermittently corrupts decode-time sparse attention output, which cascades into incoherent generation over long outputs since the model has no mechanism to recover once a wrong token is emitted from a corrupted attention result. We confirmed this is specific to the ASM kernel by swapping in an independent Triton reimplementation of the identical computation (no `num_kv_splits` concept, does the reduction differently) — full end-to-end corruption at real production scale roughly halved (2/4 → improved) and a separate deterministic-decoding (temperature=0) failure mode went from 100% reproducible to 0% once we stopped using this kernel path.

## Expected

Every call with identical inputs should return the same (correct, no NaN) result, matching a plain pure-torch reference implementation of the same computation to within normal fp32/bf16 tolerance, regardless of `num_kv_splits`.

**Comment by yhl-amd** [comment]
Repro script (`aiter_issue_repro.py`):

```python
"""
Standalone repro: aiter.mla.mla_decode_fwd (ASM sparse-MLA decode kernel)
returns NaN nondeterministically -- roughly 50% of calls -- when given
byte-identical inputs and num_kv_splits=16. num_kv_splits in {1, 8, 32} do
not reproduce this on the same inputs.

This is the decode kernel used for DeepSeek-V3.2/GLM-5.2-style sparse MLA
attention (query attends to a fixed-size top-k selected subset of the KV
cache). It is invoked with num_kv_splits=16 by at least one production
inference stack (ATOM) whenever the model is not using a "segmented" MLA
cache layout, which is the common/default case.

Environment this was found on:
    GPU:        AMD Instinct MI350-series (gfx950)
    ROCm/HIP:   7.2.53211
    torch:      2.10.0+rocm7.2.4.git3d3aa833
    aiter:      commit 95abba475 (main, 2026-07-22)
    Python:     3.12.3

Expected: every call with identical inputs returns the same (correct, no
NaN) result, matching a plain pure-torch reference implementation of the
same computation to within normal fp32/bf16 tolerance.

Actual: calling num_kv_splits in {1, 8, 16, 32} in sequence on identical
inputs -- the SWEEP below -- reproducibly (3/3 attempts on our box)
returns an output tensor that is entirely NaN (every one of
B*H*D_LATENT = 8192 elements) at num_kv_splits=16 specifically; 1, 8, and
32 are consistently correct. Calling num_kv_splits=16 alone as the very
first kernel launch in a fresh process does NOT reproduce it, and looping
num_kv_splits=16 100x in isolation reproduced it inconsistently across
runs on our box (once at 50/100, twice at 0/100) -- i.e. the failure
appears to depend on GPU state left over from preceding kernel activity,
not purely on the input tensors. This is consistent with a missing
synchronization / memory-visibility bug in the num_kv_splits=16 code
path's split-K reduction stage (a worker's partial-sum write being read
before it is guaranteed visible to the reducing thread/block, with the
race window's outcome depending on scheduling/occupancy left over from
prior GPU work). The SWEEP reproduction below is the reliable one to run
first.

Run:
    python3 aiter_issue_repro.py
"""
import torch
from aiter.mla import mla_decode_fwd

torch.manual_seed(0)
dev = "cuda"

# Real-world dimensions: DeepSeek-V3.2/GLM-5.2 MLA decode.
B, H = 1, 16                  # batch=1, 16 query heads (MQA-style KV)
D_LATENT, D_PE = 512, 64      # kv_lora_rank=512, qk_rope_head_dim=64
D = D_LATENT + D_PE           # 576
TOPK = 2048                   # sparse indexer's selected KV-position count
T = 65536                     # KV pool size (context length)

q = torch.randn(B, H, D, device=dev).bfloat16()
kv = torch.randn(T, D, device=dev).bfloat16()          # flat [T, D] KV pool
kv_buffer = kv.view(T, 1, 1, D)                         # [pages, page_size=1, nhead_kv=1, D]

sp_idx = torch.randperm(T, device=dev)[:TOPK].to(torch.int32)   # selected KV token ids
kv_indptr = torch.tensor([0, TOPK], device=dev, dtype=torch.int32)
qo_indptr = torch.tensor([0, 1], device=dev, dtype=torch.int32)  # 1 new token, decode
kv_last_page_lens = torch.ones(B, device=dev, dtype=torch.int32)
scale = D ** -0.5


def torch_reference(q, kv, sp_idx, indptr, scale):
    """Pure-torch ground truth for the same ragged sparse-MLA decode math."""
    B, H, D = q.shape
    D_LATENT = D - D_PE
    qf, kvf = q.float(), kv.float()
    out = torch.empty((B, H, D_LATENT), dtype=torch.float32, device=q.device)
    for b in range(B):
        idx = sp_idx[int(indptr[b]) : int(indptr[b + 1])].long()
        ksel = kvf[idx]                       # [TOPK, D]
        kl, kpe = ksel[:, :D_LATENT], ksel[:, D_LATENT:]
        for h in range(H):
            ql, qpe = qf[b, h, :D_LATENT], qf[b, h, D_LATENT:]
            score = (kl @ ql + kpe @ qpe) * scale
            w = torch.softmax(score, dim=-1)
            out[b, h] = (w[:, None] * kl).sum(0)
    return out


ref = torch_reference(q, kv, sp_idx, kv_indptr, scale)

print(f"aiter.mla.mla_decode_fwd nondeterminism repro  (B={B} H={H} D={D} topk={TOPK} ctx={T})")
print()

# --- Sweep num_kv_splits on identical inputs ---
for nsplits in (1, 8, 16, 32):
    o = torch.empty((B, H, D_LATENT), dtype=torch.float32, device=dev)
    mla_decode_fwd(
        q, kv_buffer, o, qo_indptr, kv_indptr, sp_idx, kv_last_page_lens,
        max_seqlen_q=1, page_size=1, sm_scale=scale, num_kv_splits=nsplits,
    )
    diff = (o.float() - ref).abs()
    print(f"  num_kv_splits={nsplits:>3}: max_abs_diff_vs_ref={float(diff.amax()):.5f} "
          f"nan_count={int(torch.isnan(o).sum())}/{o.numel()}")

print()
print("If num_kv_splits=16 above did NOT show NaN, try running the full script")
print("2-3 times -- on our box the sweep reproduced it 3/3 attempts, but exact")
print("timing-dependent bugs can vary by GPU occupancy/driver/hardware.")
print()
print("--- Repeating num_kv_splits=16 100x on the SAME inputs (supplementary; ---")
print("--- less reliable in isolation than the sweep above, see module docstring) ---")
nan_hits = 0
N = 100
for _ in range(N):
    o = torch.empty((B, H, D_LATENT), dtype=torch.float32, device=dev)
    mla_decode_fwd(
        q, kv_buffer, o, qo_indptr, kv_indptr, sp_idx, kv_last_page_lens,
        max_seqlen_q=1, page_size=1, sm_scale=scale, num_kv_splits=16,
    )
    if torch.isnan(o).any():
        nan_hits += 1
print(f"  {nan_hits}/{N} calls returned NaN despite byte-identical inputs every call.")
if nan_hits not in (0, N):
    print("  ==> NONDETERMINISTIC on identical inputs. Confirms a race condition,")
    print("      not a fixed wrong-answer-for-this-input bug.")
```

**Comment by yhl-amd** [comment]
## Additional finding: the sibling `mla_prefill_fwd` shows the same nondeterministic-NaN pattern at other `num_kv_splits` values

Tested `aiter.mla.mla_prefill_fwd` the same way (single query token, B=1, H=16, D=576, topk=2048/65536-token context, so no confound from causal masking across multiple query rows). Calling it 20x with byte-identical inputs:

```
num_kv_splits=None: nan_counts={0}   -- clean, and this is what real GLM-5.2 prefill actually calls with
num_kv_splits=1:    nan_counts={0, 12, 13, 14, 17, 20, 21, 22}  -- nondeterministic NaN
num_kv_splits=8:    nan_counts={0}   -- clean
num_kv_splits=16:   nan_counts={0, 12, 13, 14, 17, 20, 21, 22}  -- nondeterministic NaN
num_kv_splits=32:   nan_counts={0}   -- clean
```

`num_kv_splits=None` (production's actual value for prefill) is not affected in this test — confirmed byte-identical output across 10 repeated calls, max_abs_diff=0.0 every time. But `1` and `16` reproduce the same kind of nondeterministic partial-NaN pattern as the decode kernel's `num_kv_splits=16` case above, suggesting this isn't isolated to `mla_decode_fwd` — it looks like a shared defect in how the split-K reduction is synchronized across both MLA prefill and decode entry points at certain split counts, not something specific to one kernel's implementation.

Not currently a known production-impacting path (real callers don't request `num_kv_splits∈{1,16}` for prefill), but worth fixing alongside the decode case since it points at the same root cause in shared reduction code.
